Display connected peers in TUI

philipr-za opened the pull request:
## Description
This PR adds monitoring of the Connectivity Manager event stream and display of the currently connected Peers in the Network Tab of the TUI.

When a connectivity event is detected that would affect the connected peer list the list is updated in the backend.

## How Has This Been Tested?
Manually tested

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I ran `cargo test` successfully before submitting my PR.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

sdbondi commented on applications/tari_console_wallet/src/ui/state/app_state.rs:
```diff
@@ -367,13 +380,32 @@ impl AppStateInner {
+            }
+        }
+
+        self.data.connected_peers = peers;
```
nit: Single allocation, clone peer manager once

```suggestion
        let peer_manager = self.wallet.comms.peer_manager();
        let mut peers = Vec::with_capacity(connections.len());
        for c in connections {
            let peer = peer_manager.find_by_node_id(c.peer_node_id()).await?;
            peers.push(peer);
        }

        self.data.connected_peers = peers;
```

sdbondi commented on applications/tari_console_wallet/src/ui/state/app_state.rs:
```diff
@@ -195,6 +204,10 @@ impl AppState {
         }
     }
+
+    pub fn get_connected_peers(&self) -> &Vec<Peer> {
```
nit: (thought this would be a 📎 , guess not)
```suggestion
    pub fn get_connected_peers(&self) -> &[Peer] {
```

philipr-za replied:
Don't really see the benefit here? Makes it clear what collection is being stored inside and its returning a reference not a clone.

sdbondi replied:
https://rust-lang.github.io/rust-clippy/master/#ptr_arg
 The caller doesnt care that it's a vec - it just needs a view into the data. It really doesn't matter - just mentioning that you should basically never do a `&Vec<_>` anywhere according to clippy style rules, I just follow them blindly :P

philipr-za replied:
That lint is for arguments not returns. 🤷‍♂️ really don't see the benefit of hiding the fact that it's a vec.

sdbondi replied:
There is no benefit either way - it's consistency - it doesnt matter

sdbondi commented on applications/tari_console_wallet/src/ui/components/network_tab.rs:
```diff
@@ -14,10 +14,41 @@ impl NetworkTab {
+        let mut column0_items = Vec::new();
+        let mut column1_items = Vec::new();
+        let mut column2_items = Vec::new();
+        for p in peers.iter() {
```
nit:
```suggestion
        for p in peers {
```

sdbondi commented on applications/tari_console_wallet/src/ui/components/network_tab.rs:
```diff
@@ -14,10 +14,41 @@ impl NetworkTab {
+        let peers = app_state.get_connected_peers();
+        let mut column0_items = Vec::new();
+        let mut column1_items = Vec::new();
+        let mut column2_items = Vec::new();
```
nit: single allocation each

```suggestion
        let mut column0_items = Vec::with_capacity(peers.len());
        let mut column1_items = Vec::with_capacity(peers.len());
        let mut column2_items = Vec::with_capacity(peers.len());
```

sdbondi commented on applications/tari_console_wallet/src/ui/state/wallet_event_monitor.rs:
```diff
@@ -77,6 +80,26 @@ impl WalletEventMonitor {
+                                        _ => (),
+                                     }
+                            },
+                            Err(e) => error!(target: LOG_TARGET, "Error reading from Connectivity event broadcast channel"),
```
Unused var `e`? (Is clippy even working?)

nit:
An error! saying "error reading" is not an ideal way to handle this case - the error from `Stream::next` on a tokio broadcast channel happens if 1) There are no senders (disconnected)(This cannot happen because that would be the end of the stream and this block would not execute), or 2) this loop could not read off the channel fast enough and some messages were dropped (lagged).
In this case, ignoring the error is the best option unless you can do something in response to missing some messages.

sdbondi replied:
So the only thing you can do is log that there is lagging happening - which could of course help make a slow select loop more noticeable in the logs. I'd make it debug and mention lagging, or not log it and make sure my loop isn't excessively slow.

sdbondi reviewed (commented):
LGTM - some nits

sdbondi commented on applications/tari_console_wallet/src/ui/state/wallet_event_monitor.rs:
```diff
@@ -74,7 +77,27 @@ impl WalletEventMonitor {
+                                        _ => (),
+                                     }
+                            },
+                            Err(e) => debug!(target: LOG_TARGET, "Lagging read on Connectivity event broadcast channel"),
```
Unused `e`

sdbondi reviewed (dismissed):
LGTM

sdbondi commented on applications/tari_console_wallet/src/ui/state/app_state.rs:
```diff
@@ -367,13 +380,33 @@ impl AppStateInner {
         self.wallet.transaction_service.get_event_stream_fused()
     }
+
+    pub async fn get_connectivity_event_stream(&self) -> Fuse<ConnectivityEventRx> {
```
async not needed
```suggestion
    pub fn get_connectivity_event_stream(&self) -> Fuse<ConnectivityEventRx> {
```

philipr-za committed:
Display connected peers in TUI

This PR adds monitoring of the Connectivity	Manager event stream and display of the currently connected Peers in the Network Tab of the TUI.

When a connectivity event is detected that would affect the connected peer list the list is updated in the backend.
